include/ctpg/ctpg.hpp

```diff
@@ -1638,7 +1644,61 @@ namespace detail
     constexpr size_t value_stack_initial_capacity = 1 << 10;
     constexpr size_t cursor_stack_initial_capacity = 1 << 10;
 
-    template<typename CursorStack, typename ValueStack, typename ErrorStream, typename Iterator>
+    template<typename Context, typename ValueVariantType, typename RuleTupleType, size_t RuleCount>
+    struct value_reductors
+    {
+        constexpr value_reductors(Context &&context, const RuleTupleType& rule_tuple): context(context), rule_tuple(rule_tuple) {
```

Comment:
shouldn't the context be forwarded here? std::forward<> ?

Reply:
also, if the value_reductors object is owning the context, I think it should be stored as std::decay_t<Context>